### ISSUE TITLE
Graph partitioning

### ISSUE THREAD

**Issue by nassimGg** [facebookresearch/PyTorch-BigGraph#41]
Hello!
Thank you for this amazing project.
I'd like to ask if we can consider the embedding of each Partition as an embedding of sub-graph?

Regards,

**Comment by lerks** [comment]
I don't think there's a clear answer to your question. It depends on what you mean by "consider".

For example, you could train embeddings for a big graph and, once finished, extract the embeddings of a subset of entities (it could be an entire partition, more, less, or any subset really). This subset of embeddings will allow you to predict the scores of edges between entities in this subset, and these scores will be the same as if you calculated them with the full model.

What I just said can be rephrased as: after training, you can drop as many entities as you want and the predictions you can make over the entities that remain will still maintain the same quality.

However, _training_ using a single partition of entities will definitely produce lower-quality embeddings than if you trained on a larger graph and then threw away the entities you are not interested in.

For example, consider this graph with 6 vertices: https://i.stack.imgur.com/2IYNr.png

If your partition happens to be {A, B, C} and {M, N, P} and you wanted to just know the embeddings of {A, B, C} you cannot train just on the {A, B, C} partition because there are no edges among these 3 vertices. You would have to train on the entire graph (which allows you to learn something) and then forget that {M, N, P} ever existed.

Hope I didn't grossly misunderstand the question and that these pointers can prove helpful.

**Comment by nassimGg** [comment]
Hello @ledw,
Thank you very much for your quick and apt response.
Just another question is that if we can use PyTorchBigGraph as Graph2Vec Embeddings note Node2Vec Embeddings.

Regars,

**Comment by lerks** [comment]
Sorry for the late response, this slipped through. I am not very familiar with graph2vec and node2vec so I just read up on them.

Graph2vec is a system to learn an embedding for the whole graph, not for each of its vertices, so it achieves a different purpose than PBG and I wouldn't know how to naturally compare them.

Node2vec is more in the same spirit as PBG. I believe PBG could to some extent replicate node2vec, at least in some limited setting (their loss function should be implementable in PBG, although I didn't check thoroughly). However, one aspect on which they put a lot of weight is the neighborhood sampling, because they consider more than the immediate neighbors. This is almost impossible to replicate in PBG because for performance reason required by scalability we use a very specific sampling technique which is deeply baked in and hard to replace.

In my experience, negative sampling is the most critical aspect of systems like these, on which most of the performance depends on. The science also isn't necessarily all in yet, so it is definitely an open and interesting research question how much of an advantage sampling larger neighborhoods brings (if any) and in which cases it's more important. Let me point out that DeepWalk also uses a sophisticated negative sampling scheme but, in the experiments in [our paper](https://www.sysml.cc/doc/2019/71.pdf), we've seen that PBG performs better, at least in the cases and metrics we considered. I don't have a similar comparison for node2vec, but it would definitely be interesting to investigate it.

Let me also conclude by saying that PBG was designed for large scales and to get there it had to accept some compromises. While a comparison with other methods on accuracy is interesting and necessary, it may be unfair to PBG if it doesn't also take into account performances on big graphs.
